lib/apivore/validator.rb

```diff
@@ -0,0 +1,116 @@
+require 'action_controller'
+require 'action_dispatch'
+
+module Apivore
+  class Validator
+    include ::ActionDispatch::Integration::Runner
+
+    attr_reader :method, :path, :expected_response_code, :params
+
+    def initialize(method, path, expected_response_code, params = {})
+      @method = method.to_s
+      @path = path.to_s
+      @params = params
+      @expected_response_code = expected_response_code.to_i
+    end
+
+    def matches?(swagger_checker)
+      pre_checks(swagger_checker)
+
+      unless has_errors?
+        send(
+          method,
+          full_path(swagger_checker),
+          params['_data'] || {},
+          params['_headers'] || {}
+        )
+        post_checks(swagger_checker)
+        swagger_checker.remove_tested_end_point_response(path, method, expected_response_code)
+      end
+
+      !has_errors?
+    end
+
+    def full_path(swagger_checker)
+      apivore_build_path(swagger_checker.base_path + path, params)
+    end
+
+    def apivore_build_path(path, data)
+      path.scan(/\{([^\}]*)\}/).each do |param|
+        key = param.first
+        if data && data[key]
+          path = path.gsub "{#{key}}", data[key].to_s
+        else
+          raise URI::InvalidURIError, "No substitution data found for {#{key}} to test the path #{path}.\nAdd it via an:\n  apivore_setup '<path>', '<method>', '<response>' do\n    { '#{key}' => <value> }\n  end\nblock in your specs.", caller
+        end
+      end
+      path + (data['_query_string'] ? "?#{data['_query_string']}" : '')
+    end
+
+
+    def pre_checks(swagger_checker)
+      check_request_path(swagger_checker)
+    end
+
+    def post_checks(swagger_checker)
+      check_status_code
+      check_response_is_valid(swagger_checker) unless has_errors?
+    end
+
+    def check_request_path(swagger_checker)
```

Comment:
It's great to see these methods split out like this, will make our life easier to track down errors